### PR TITLE
Added the concept of a record container with an ageoff method.  Can also

### PR DESCRIPTION
**Pull request by ivakegg**
configure the ingest helper to add an ageoff method if not already provided.

### CODE REVIEW

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/AgeOffIngest.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+public interface AgeOffIngest {
+
+    /**
+     * Get the ageoff date from the record and fields
+     *
+     * @param value
+     * @param fields
+     * @return the age off date (usually the record date)
+     */
+    default long getAgeOffDate(RawRecordContainer value, Multimap<String,NormalizedContentInterface> fields) {
```
Comment (ivakegg) [inline warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/AgeOffIngest.java]: Should I simply add this to the IngestHelperInterface ?

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/BaseIngestHelper.java [hunk 2]
```diff
@@ -171,8 +178,37 @@ public abstract class BaseIngestHelper extends AbstractIngestHelper implements C
...
+        }
+        // otherwise lets try and pull it from one of the fields
+        long ageOffDate = record.getDate();
+        if (ageoffDateField != null && !ageoffDateField.isBlank()) {
+            Collection<NormalizedContentInterface> values = fields.get(ageoffDateField);
+            if (values != null && !values.isEmpty()) {
+                // take the largest date
+                for (NormalizedContentInterface value : values) {
+                    try {
+                        Date d = new DateNormalizer().denormalize(value.getIndexedFieldValue());
```
Comment (ivakegg) [inline warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/BaseIngestHelper.java]: Should I configure a separate date format like we do for the event date in the AbstractEventRecordReader ?

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/AgeOffIngest.java (hunk 1 above)
Comment (hlgp) [reply warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/AgeOffIngest.java]: I think that's a reasonable idea

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java [hunk 3]
```diff
@@ -852,9 +881,11 @@ protected Key createKey(byte[] row, Text colf, Text colq, byte[] vis, long ts, b
      */
     protected Key createIndexKey(byte[] row, Text colf, Text colq, byte[] vis, long ts, boolean delete) {
         // Truncate the timestamp to the day
-        long tsToDay = (ts / MS_PER_DAY) * MS_PER_DAY;
+        long tsToDay = (CompositeTimestamp.getEventDate(ts) / MS_PER_DAY) * MS_PER_DAY;
+        long ageOffToDay = (CompositeTimestamp.getAgeOffDate(ts) / MS_PER_DAY) * MS_PER_DAY;
+        long indexComposite = CompositeTimestamp.getCompositeTimeStamp(tsToDay, ageOffToDay);
```
Comment (hlgp) [inline warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: perfect

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/BaseIngestHelper.java (hunk 2 above)
Comment (hlgp) [reply warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/BaseIngestHelper.java]: not sure that's necessary

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java [hunk 4]
```diff
@@ -852,9 +881,11 @@ protected Key createKey(byte[] row, Text colf, Text colq, byte[] vis, long ts, b
      */
     protected Key createIndexKey(byte[] row, Text colf, Text colq, byte[] vis, long ts, boolean delete) {
         // Truncate the timestamp to the day
-        long tsToDay = (ts / MS_PER_DAY) * MS_PER_DAY;
+        long tsToDay = (CompositeTimestamp.getEventDate(ts) / MS_PER_DAY) * MS_PER_DAY;
```
Comment (hlgp) [inline warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: worth making a truncateToDay method?

Comment (ivakegg) [reply warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: done

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/AgeOffIngest.java (hunk 1 above)
Comment (ivakegg) [reply warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/AgeOffIngest.java]: moved to raw record container

File: warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/BaseIngestHelper.java (hunk 2 above)
Comment (ivakegg) [reply warehouse/ingest-core/src/main/java/datawave/ingest/data/config/ingest/BaseIngestHelper.java]: OBE

**Review by jack-gitdev — APPROVED** [review]
Seems pretty straightforward. Looks good.

File: warehouse/core/src/test/java/datawave/util/CompositeTimestampTest.java [hunk 5]
```diff
@@ -200,4 +206,166 @@ public void testPostDoomsday() {
...
+
+        // test the same negative eventdate, but different ageoff dates
+        t1 = CompositeTimestamp.getCompositeTimeStamp(aDate, aDate);
+        t2 = CompositeTimestamp.getCompositeTimeStamp(aDate, aMonthLater);
+        t3 = CompositeTimestamp.getCompositeTimeStamp(aDate, twoMonthsLater);
+
+        // in this case the natural ordering will be incorrect ( and in fact exactly opposite )
+        Assert.assertFalse(isOrdered(t1, t2, t3));
+        Assert.assertTrue(isOrdered(t2, t2, t1));
+        // but the comparator will maintain the correct ordering
```
Comment (hlgp) [inline warehouse/core/src/test/java/datawave/util/CompositeTimestampTest.java:289]: hooray!

File: warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java [hunk 6]
```diff
@@ -737,20 +744,19 @@ public void testRemoveAndReAdd() throws Exception {
         String uuid1 = UUID.randomUUID().toString();
         ArrayList<Value> values = Lists.newArrayList();
 
-        // When we're considering timestamps, an add of a UID, followed by a removal
-        // of that UID, and then a re-add should result in the UID ending up in the
-        // UID list.
+        // an add of a UID, followed by a removal of that UID, and then a re-add
+        // should still result in the UID removed
```
Comment (hlgp) [inline warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java:748]: Why did we decide to change this behavior?

**Review by hlgp — APPROVED** [review]
Definitely fine merging it into the other feature branch.  Want to run through a few scenarios in integration testing.

**Review by keith-ratcliffe — APPROVED** [review]

File: warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java (hunk 6 above)
Comment (ivakegg) [reply warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java:748]: I removed the setTimestampsIgnoredOpt which defaults to true.  Setting it to false is actually dangerous and wrong in the case of composite timestamps.  We are making the decision that different timestamps in the global index actually mean something different (not actually versions but need to be treated separately) so this flag is not usefull anyway.

**Pull request merged**